BridgeCareApp/BridgeCare/DataAccessLayer/DeficientReport.cs

```diff
@@ -0,0 +1,104 @@
+﻿using BridgeCare.ApplicationLog;
+using BridgeCare.Interfaces;
+using BridgeCare.Models;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.SqlClient;
+using System.Linq;
+
+namespace BridgeCare.DataAccessLayer
+{
+    public class DeficientReport : IDeficientReport
+    {
+        private readonly BridgeCareContext db;
+        private readonly TargetsMet getDeficients;
+        private readonly CellAddress address;
+
+        public DeficientReport(BridgeCareContext context, TargetsMet deficients, CellAddress cell)
+        {
+            db = context ?? throw new ArgumentNullException(nameof(context));
+            getDeficients = deficients ?? throw new ArgumentNullException(nameof(deficients));
+            address = cell ?? throw new ArgumentNullException(nameof(cell));
+        }
+        public DeficientResult GetData(SimulationModel data, int[] totalYears)
+        {
+            // Deficient and DeficientResults are models. Deficient gets data from the database.
+            // DeficientResult gets the processed data
+            IQueryable<DeficientModel> deficients = null;
+            DeficientResult result = null;
+
+            var select =
+                "SELECT TargetID, Years, TargetMet, IsDeficient " +
+                    " FROM Target_" + data.NetworkId
+                    + "_" + data.SimulationId;
+
+            try
+            {
+                var rawDeficientList = db.Database.SqlQuery<DeficientModel>(select).AsQueryable();
+
+                deficients = rawDeficientList.Where(_ => _.IsDeficient == true);
+
+                var targetAndYear = getDeficients.GetData(deficients);
+                result = GetDeficientInformation(data, targetAndYear, totalYears);
+            }
+            catch (SqlException ex)
+            {
+                HandleException.SqlError(ex, "Target");
+            }
+            catch (OutOfMemoryException ex)
+            {
+                HandleException.OutOfMemoryError(ex);
+            }
+            catch (Exception ex)
+            {
+                HandleException.GeneralError(ex);
+            }
+            return result;
+        }
+
+        private DeficientResult GetDeficientInformation(SimulationModel data, Hashtable YearsIDValues, int[] totalYears)
+        {
+            var deficientTableData = db.Deficient.AsNoTracking().Where(_ => _.SimulationID == data.SimulationId);
+            var totalYearCount = totalYears.Count();
+            DataTable DeficientTable = new DataTable();
+            DeficientTable.Columns.Add("Attribute");
+            DeficientTable.Columns.Add("Group");
+            for (int i = 0; i < totalYearCount; i++)
+            {
+                DeficientTable.Columns.Add(totalYears[i].ToString());
+            }
+            var increment = 2;
+            foreach (var item in deficientTableData)
```

Comment:
If possible, extract this looping behavior out to an object.

Reply:
Extracted it out in private method